### ISSUE TITLE
Significant degradation of multi-stage Interleaved1F1B performance on MoE models

### ISSUE THREAD

**Issue by yuankaichen-amd** [pytorch/torchtitan#2209]
### Bug description

Model: deepseek-v3-16B with pipeline enabled (in my testing, I changed the number of layers to 2) :

pipeline_parallel_degree = 2
pipeline_parallel_first_stage_less_layers = 0
pipeline_parallel_last_stage_less_layers = 0
pipeline_parallel_layers_per_stage = 2

We observe that the next stage's recv (e.g. PP:2Recv_F0) gets scheduled before the current stage (PP:0F1)

<img width="2183" height="424" alt="Image" src="https://github.com/user-attachments/assets/9c625d64-af28-4af9-89d9-4c089aab3a2e" />

For MoE models, PP:0F1 contains all2all comm and thus PP:2Recv_F0 blocks the all2all:

<img width="982" height="468" alt="Image" src="https://github.com/user-attachments/assets/045b5942-b53c-41e5-b1f0-e8f023e5587c" />

All2all cannot start util the current rank has received the data for its next stage, and each time it injects a bubble into the pipeline. The pipeline is now full of bubbles:

<img width="2606" height="494" alt="Image" src="https://github.com/user-attachments/assets/95926085-75c0-48fb-bf68-75a9deafe5cb" />

### Versions

latest Pytorch-rocm-7.1 nightly, latest torchtitan

**Comment by sanketpurandare** [comment]
Hi @yuankaichen-amd if the total layers in your model are 2, pipeline_parallel_layers_per_stage = 2, pipeline_parallel_degree = 2. The interleaved 1F1B schedule requires at least 2 stages per pp_rank as shown below:

<img width="1730" height="721" alt="Image" src="https://github.com/user-attachments/assets/349758c2-f714-403e-914c-e1901101787d" />

 How are you managing the layer to stage assignment?

Second, coming to your question of pp collective scheduling, i tried the following utility:

```
from torch.distributed.pipelining._schedule_visualizer import get_schedule_ops

schedule_ops = get_schedule_ops(
    schedule="Interleaved1F1B", pp_degree=2, num_microbatches=4, with_comms=True
)
for row in schedule_ops:
    print(row)
```

Output:

```
[0UNSHARD, 2UNSHARD, 0F0, 0SEND_F0, 2RECV_F0, 0F1, 0SEND_F1, 2RECV_F1, 2F0, 2SEND_F0, 2F1, 2SEND_F1, 2RECV_B0, 2B0, 2SEND_B0, 2RECV_B1, 2B1, 2SEND_B1, 2REDUCE_GRAD, 2RESHARD, 0RECV_B0, 0B0, 0RECV_B1, 0B1, 0REDUCE_GRAD, 0RESHARD]
[1UNSHARD, 3UNSHARD, 1RECV_F0, 1F0, 1SEND_F0, 1RECV_F1, 1F1, 1SEND_F1, 3RECV_F0, 3F0, 3RECV_F1, 3B0, 3SEND_B0, 1RECV_B0, 3F1, 3B1, 3SEND_B1, 1RECV_B1, 3REDUCE_GRAD, 3RESHARD, 1B0, 1SEND_B0, 1B1, 1SEND_B1, 1REDUCE_GRAD, 1RESHARD]
```

You are right that 2RECV_F0 is scheduled before 0F1 on pp_rank 0, This is because the order of collectives on pp_rank 1 is 1SEND_F0, 1RECV_F1. 

For 0F1:

Stage 0 is is_first=True
So it skips the wait entirely and runs the forward immediately
0F1 runs on the compute stream and the all2all runs on a different stream than the pp comms, specifically on the EP NCCL stream.
2RECV_F0 is posted on the PP NCCL stream asynchronously
They are on different streams and there's no synchronization between them until 2F0 runs and explicitly waits on 2RECV_F0.

So:

2RECV_F0 posts async recv on PP stream (non-blocking)
0F1 runs on compute stream/EP NCCL stream (independent, not blocked)
Any non-PP collectives inside 0F1 (e.g., all2all on a different process group) run on their own streams and are also not blocked
Only 2F0 will wait on 2RECV_F0
The scheduling of 2RECV_F0 before 0F1 has no performance impact on 0F1 execution.

**Comment by ych622** [comment]
Hi @sanketpurandare ,  thanks for looking into this!   I made a mistake in the description -- the number of layers was 8 (not 2), so each rank has 2 stages and each stage has 2 layers.

"2RECV_F0 posts async recv on PP stream (non-blocking)"

I thought it was the case as well -- the trace does confirm that pp and all2all are on separate streams, but for some reason pp blocks the all2all in our case.

I'm wondering if you have a working trace on Nvidia hardware? Then we can narrow it down to our collective library issue.

**Comment by ych622** [comment]
@H-Huang   @sanketpurandare   In my old trace (3 month ago), the pp send/recv never overlaps with F/B

<img width="2300" height="342" alt="Image" src="https://github.com/user-attachments/assets/5b5fa8ab-5286-4f32-997a-55341c01cc6d" />

Also there was only one batched send/recv between stages

<img width="894" height="484" alt="Image" src="https://github.com/user-attachments/assets/ceeb0973-534a-4426-9f91-150c5900e557" />

The current trace shows two consecutive send and recv.

Do you know what change may have caused this?

**Comment by H-Huang** [comment]
Looking into this, let me get a trace with those settings and share it here

**Comment by H-Huang** [comment]
Command:

`NGPU=4 CONFIG_FILE=/data/users/howardhuang/titan2/torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml ./run_train.sh`

Here is my diff for the config:

<details>
diff --git a/torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml b/torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml
index 00ec5331..d5d3cd41 100644
--- a/torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml
+++ b/torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml
@@ -4,14 +4,14 @@ description = "DeepSeek-V3 16B model training"
 print_config = false
 
 [profiling]
-enable_profiling = false
+enable_profiling = true
 save_traces_folder = "profile_trace"
-profile_freq = 10
+profile_freq = 4
 enable_memory_snapshot = false
 save_memory_snapshot_folder = "memory_snapshot"
 
 [metrics]
-log_freq = 10
+log_freq = 1
 disable_color_printing = false
 enable_tensorboard = false
 save_tb_folder = "tb"
@@ -29,7 +29,7 @@ lr = 2.2e-4
 eps = 1e-8
 
 [lr_scheduler]
-warmup_steps = 200  # lr scheduler warm up, normally 20% of the train steps
+warmup_steps = 3  # lr scheduler warm up, normally 20% of the train steps
 decay_ratio = 0.8  # lr scheduler decay ratio, 80% of the train steps
 decay_type = "cosine"
 min_lr_factor = 0.1
@@ -38,7 +38,7 @@ min_lr_factor = 0.1
 local_batch_size = 4
 seq_len = 4096
 max_norm = 1.0  # grad norm clipping
-steps = 1000
+steps = 10
 dataset = "c4"  # supported datasets: c4_test (2K), c4 (177M)
 
 [parallelism]
@@ -47,10 +47,13 @@ data_parallel_shard_degree = -1
 fsdp_reshard_after_forward = "default" # default / never / always
 tensor_parallel_degree = 1
 enable_async_tensor_parallel = false
-pipeline_parallel_degree = 1
+pipeline_parallel_degree = 2
 pipeline_parallel_schedule = "Interleaved1F1B"
-expert_parallel_degree = 8
+expert_parallel_degree = 2
 expert_tensor_parallel_degree = 1
+pipeline_parallel_first_stage_less_layers = 0
+pipeline_parallel_last_stage_less_layers = 0
+pipeline_parallel_layers_per_stage = 2
 
 [checkpoint]
 enable = false
</details>

For the most part I see overlap as expected: 

<img width="1565" height="777" alt="Image" src="https://github.com/user-attachments/assets/ac5683fa-a205-4ec6-b718-ef8b1d8f8936" />

Am I missing anything?

Attached my traces as well

[iteration_8.zip](https://github.com/user-attachments/files/24626462/iteration_8.zip)

**Comment by yuankaichen-amd** [comment]
Thanks @H-Huang !  Your trace looks working exactly to the design. So here we need to find out why -

(1) Our pp comm is blocking other comms (we will look into this);
(2) The pp schedule seems different against our earlier trace. (let's discuss this offline, DM'ed you on slack)

**Comment by yuankaichen-amd** [comment]
Thanks @H-Huang for the very helpful inputs!  Now we understand the pp schedule diff.

We filed https://github.com/pytorch/pytorch/issues/172668 to track this issue from the pytorch side.
